src/txmempool.cpp

```diff
@@ -813,17 +815,23 @@ CTransactionRef CTxMemPool::get(const uint256& hash) const
 
 TxMempoolInfo CTxMemPool::info(const GenTxid& gtxid) const
 {
+    AssertLockNotHeld(cs);
     LOCK(cs);
     indexed_transaction_set::const_iterator i = (gtxid.IsWtxid() ? get_iter_from_wtxid(gtxid.GetHash()) : mapTx.find(gtxid.GetHash()));
     if (i == mapTx.end())
         return TxMempoolInfo();
     return GetInfo(i);
 }
 
-TxMempoolInfo CTxMemPool::info(const uint256& txid) const { return info(GenTxid{false, txid}); }
+TxMempoolInfo CTxMemPool::info(const uint256& txid) const
+{
+    AssertLockNotHeld(cs);
```

Comment:
Excluding locks in getters makes it impossible to get more than one consistent mempool entry, as the state of the second might change after the first one has been returned.

Reply:
`AssertLockNotHeld` is called immediately in `info(gtxid)` seems unnecessary to repeat it?

We've got `infoAll` for a consistent look at all mempool entries (used by DumpMempool), and apparently don't use `info` for consistent views otherwise, so excluding the lock seems okay to me.

Reply:
To get consistent values the external locking of `CTxMemPool::cs` should be used.

Anyway, this PR does not change locking behavior.

Reply:
> `AssertLockNotHeld` is called immediately in `info(gtxid)` seems unnecessary to repeat it?

When a thread safety annotation is a part of function member declaration, I found it very convenient having a corresponding `AssertLock{Not}Held()` statement in its definition.